crates/bevy_core_pipeline/src/dof/mod.rs

```diff
@@ -0,0 +1,907 @@
+    pipeline_cache: Res<PipelineCache>,
+    mut pipelines: ResMut<SpecializedRenderPipelines<DepthOfFieldPipeline>>,
+    msaa: Res<Msaa>,
+    global_bind_group_layout: Res<DepthOfFieldGlobalBindGroupLayout>,
+    view_targets: Query<(
+        Entity,
+        &ExtractedView,
+        &DepthOfFieldSettings,
+        &ViewDepthOfFieldBindGroupLayouts,
+    )>,
+) {
+    for (entity, view, dof_settings, view_bind_group_layouts) in view_targets.iter() {
+        let dof_pipeline = DepthOfFieldPipeline {
+            view_bind_group_layouts: view_bind_group_layouts.clone(),
+            global_bind_group_layout: global_bind_group_layout.layout.clone(),
+        };
+
+        // We'll need these two flags to create the `DepthOfFieldPipelineKey`s.
+        let (hdr, multisample) = (view.hdr, *msaa != Msaa::Off);
+
+        // Go ahead and specialize the pipelines.
+        match dof_settings.mode {
+            DepthOfFieldMode::Gaussian => {
+                commands
+                    .entity(entity)
+                    .insert(DepthOfFieldPipelines::Gaussian {
+                        horizontal: pipelines.specialize(
+                            &pipeline_cache,
+                            &dof_pipeline,
+                            DepthOfFieldPipelineKey {
+                                hdr,
+                                multisample,
+                                pass: DofPass::GaussianHorizontal,
+                            },
+                        ),
+                        vertical: pipelines.specialize(
+                            &pipeline_cache,
+                            &dof_pipeline,
+                            DepthOfFieldPipelineKey {
+                                hdr,
+                                multisample,
+                                pass: DofPass::GaussianVertical,
+                            },
+                        ),
+                    });
+            }
+
+            DepthOfFieldMode::Bokeh => {
+                commands
+                    .entity(entity)
+                    .insert(DepthOfFieldPipelines::Bokeh {
+                        pass_0: pipelines.specialize(
+                            &pipeline_cache,
+                            &dof_pipeline,
+                            DepthOfFieldPipelineKey {
+                                hdr,
+                                multisample,
+                                pass: DofPass::BokehPass0,
+                            },
+                        ),
+                        pass_1: pipelines.specialize(
+                            &pipeline_cache,
+                            &dof_pipeline,
+                            DepthOfFieldPipelineKey {
+                                hdr,
+                                multisample,
+                                pass: DofPass::BokehPass1,
+                            },
+                        ),
+                    });
+            }
+        }
```

Comment:
Up to you if you feel this is worthwhile, but you could do this to reduce a bit of nesting.

```suggestion
        // Go ahead and specialize the pipelines.
        let pipelines = match dof_settings.mode {
            DepthOfFieldMode::Gaussian => DepthOfFieldPipelines::Gaussian {
                horizontal: pipelines.specialize(
                    &pipeline_cache,
                    &dof_pipeline,
                    DepthOfFieldPipelineKey {
                        hdr,
                        multisample,
                        pass: DofPass::GaussianHorizontal,
                    },
                ),
                vertical: pipelines.specialize(
                    &pipeline_cache,
                    &dof_pipeline,
                    DepthOfFieldPipelineKey {
                        hdr,
                        multisample,
                        pass: DofPass::GaussianVertical,
                    },
                ),
            },
            DepthOfFieldMode::Bokeh => DepthOfFieldPipelines::Bokeh {
                pass_0: pipelines.specialize(
                    &pipeline_cache,
                    &dof_pipeline,
                    DepthOfFieldPipelineKey {
                        hdr,
                        multisample,
                        pass: DofPass::BokehPass0,
                    },
                ),
                pass_1: pipelines.specialize(
                    &pipeline_cache,
                    &dof_pipeline,
                    DepthOfFieldPipelineKey {
                        hdr,
                        multisample,
                        pass: DofPass::BokehPass1,
                    },
                ),
            },
        };
        commands.entity(entity).insert(pipelines);
```